lucene/sandbox/src/java/org/apache/lucene/search/LargeNumHitsTopDocsCollector.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.search;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+
+import org.apache.lucene.index.LeafReaderContext;
+
+import static org.apache.lucene.search.TopDocsCollector.EMPTY_TOPDOCS;
+
+/**
+ * Optimized collector for large number of hits.
+ * The collector maintains an ArrayList of hits until it accumulates
+ * the requested number of hits. Post that, it builds a Priority Queue
+ * and starts filtering further hits based on the minimum competitive
+ * score.
+ */
+public final class LargeNumHitsTopDocsCollector implements Collector {
+  private final int numHits;
+  private List<ScoreDoc> hits = new ArrayList<>();
+  /** Whether {@link #totalHits} is exact or a lower bound. */
+  private TotalHits.Relation totalHitsRelation = TotalHits.Relation.EQUAL_TO;
+  // package private for testing
+  HitQueue pq;
+  ScoreDoc pqTop;
+  int totalHits;
+
+  public LargeNumHitsTopDocsCollector(int numHits) {
+    this.numHits = numHits;
```

Comment:
Can we rename `numHits` to `requestedHitCount` or so?  Otherwise `numHits` and `totalHits` are confusable?